torch_xla/csrc/xla_graph_executor.cpp

```diff
@@ -468,10 +468,37 @@ void XLAGraphExecutor::MaybeDumpGraph(std::string name,
   }
 }
 
-XLAGraphExecutor::ComputationCache* XLAGraphExecutor::GetComputationCache() {
+XLAGraphExecutor::ComputationCache* createComputationCache() {
   static const size_t kMaxCacheSize =
       runtime::sys_util::GetEnvInt("XLA_COMPILATION_CACHE_SIZE", 1024);
-  static ComputationCache* cache = new ComputationCache(kMaxCacheSize);
+  static std::string persistentCacheDir = runtime::sys_util::GetEnvString(
+      "XLA_PERSISTENT_COMPILATION_CACHE_PATH", "");
+  if (persistentCacheDir != "") {
+    auto serialize_fn = [](auto& computation, auto& ostream) -> bool {
+      return runtime::GetComputationClient()->SerializeComputation(
+          computation->computation->client_computation(), ostream);
+    };
+    auto deserialize_fn = [](auto& istream)
+        -> std::shared_ptr<XLAGraphExecutor::CachedComputation> {
+      auto client_computation =
+          runtime::GetComputationClient()->DeserializeComputation(istream);
+      if (!client_computation) return nullptr;
+      auto computation =
+          std::make_shared<torch_xla::Computation>(client_computation);
+      return std::make_shared<XLAGraphExecutor::CachedComputation>(
+          computation, /*is_sharded=*/true);
+    };
+    /* In a distributed SPMD context, only one worker should write. */
+    bool readonly = runtime::GetComputationClient()->GetProcessIndex() != 0;
```

Comment:
I'm not sure this is right in general. It's easy for an ordinal to sneak into the graph as a constant with multiprocessing, particularly when constructing a mask for e.g. pinned all_gather. I believe each replica will need its own cache.

What if you always write to `$XLA_PERSISTENT_COMPILATION_CACHE_PATH/$ORDINAL`? This should have no impact on the SPMD case. You'll just write to `$XLA_PERSISTENT_COMPILATION_CACHE_PATH/0`.

Reply:
This PR only supports SPMD, so we'll need to follow up with MP. This is an optimization for a shared mount, since thousands of workers writing to the cache at once could cause some issues.

In MP, would we need a different cache for each worker? Any difference in the graph should be reflected in the hash, right?

Reply:
I'll make `readonly` controlled by a flag. It won't always be a shared mount even for SPMD.

Reply:
I updated to read `readonly` from an environment variable.